Refactoring of NearestNeighborMapping Class and Added Nearest Neighbor Gradient Mapping

ariguiba opened the pull request:
New Mapping, Nearest Neighbor Gradient, is added and Refactoring of Nearest Neighbor and Nearest Neighbor Gradient to inherit from one common parent

## Main changes of this PR

Nearest Neighbor Gradient mapping is a new mapping method that uses the same base principle as Nearest Neighbor, by mapping the vertices from the mesh to the nearest one in the other mesh. It tries to reduce the error, by additionally using the gradient data of the vertices to get a more accurate approximation of the output value.

The Nearest Neighbor Mapping class is changed to a Nearest Neighbor Base Mapping class that contains the same methods, except for the value mapping method. This class serves as base for the 2 different kinds of mappings, since the computations of the vertices is unchanged. 

## Motivation and additional information

To reduce code duplicates between the different mapping methods, a base class for both Nearest Neighbor and Nearest Neighbor Gradient is created, and so only the value-mapping has to be implemented (and eventually improved for efficiency) in the children classes.

## Author's checklist

* [x] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [x] I ran `make format` to ensure everything is formatted correctly.
* [x] I sticked to C++14 features.
* [x] I sticked to CMake version 3.10.
* [x] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [x] Does the changelog entry make sense? Is it formatted correctly?
* [x] Do you understand the code changes?

DavidSCN commented:
Did you intentionally leave out the tests in the PR? The headline of your PR states that it is actually only a refactoring, but you still add new methods, right? If you want to add the new methods, you also need to add the corresponding tests.

ariguiba commented:
> Did you intentionally leave out the tests in the PR? The headline of your PR states that it is actually only a refactoring, but you still add new methods, right? If you want to add the new methods, you also need to add the corresponding tests.

Yes the tests were left out intentionally because they require some added methods to the mesh classes and the corresponding configuration classes. Those classes were not included because otherwise it would be too confusing. But I imagine I can add those, as well as the corresponding tests.

DavidSCN commented on src/mesh/Data.cpp:
```diff
@@ -13,7 +14,9 @@ size_t Data::_dataCount = 0;
     : _name(""),
       _id(-1),
-      _dimensions(0)
+      _dimensions(0),
+      _meshDimensions(-1),
+      _hasGradient(false)
```
You have quite some repetition in the constructors here and I think you should use a constructor delegation. In principle, you can also assign default values within the class declaration.

DavidSCN commented on src/mesh/Data.hpp:
```diff
@@ -83,7 +100,13 @@ class Data {
+  /// Returns if the data contains gradient data
+  bool hasGradient() const;
+
+  /// Returns the mesh dimension (i.e., number of rows) of one gradient data value .
```
Isn't the mesh dimension the same as the overall spatial dimension? Maybe the words 'spatial' and 'data' help to distinguish here.

DavidSCN commented on src/mesh/Data.hpp:
```diff
@@ -94,14 +117,22 @@ class Data {
   /// ID of the data set (supposed to be unique).
   DataID _id;
 
+  /// Dimensionality of one mesh elements -> number of rows (only 2, 3 allowed for 2D, 3D).
```
```suggestion
  /// Dimension of one mesh element -> number of rows (only 2, 3 allowed for 2D, 3D).
```

DavidSCN commented on src/mesh/Data.hpp:
```diff
@@ -94,14 +117,22 @@ class Data {
   /// Dimensionality of one data value.
   int _dimensions;
+
+  /// Flag is the gradient data is available
```
```suggestion
  /// Flag if the gradient data is available
```

DavidSCN reviewed (commented):
Looks good overall, only some minor comments. I know, some of them are probably referring to legacy code, but I would fix them now.

ariguiba commented on src/mapping/Mapping.cpp:
```diff
@@ -127,8 +127,10 @@ bool operator<(Mapping::MeshRequirement lhs, Mapping::MeshRequirement rhs)
-    return rhs == Mapping::MeshRequirement::FULL;
+    return rhs == Mapping::MeshRequirement::FULL || rhs == Mapping::MeshRequirement:: GRADIENT;
   case (Mapping::MeshRequirement::FULL):
+    return rhs == Mapping::MeshRequirement::GRADIENT;
```
I assumed that the order in this case is not very relevant, it needs to be mainly consistent everywhere in the code. For now the enums are given an integer value (1 for VERTEX, 2 for FULL and 3 for GRADIENT) so this was the order I used. But indeed, it makes sense to assume the FULL Requirement would be ranked higher.  Should I also change their respective integer values?

DavidSCN replied:
> I assumed that the order in this case is not very relevant, it needs to be mainly consistent everywhere in the code.

No, the order and integer values are not relevant (and you don't need to change it). I just said 'lower' since its actually the 'lower' operator you define here (`operator<`).

ariguiba commented on src/mapping/NearestNeighborBaseMapping.cpp:
```diff
@@ -0,0 +1,153 @@
+void NearestNeighborBaseMapping::tagMeshSecondRound()
+{
+  PRECICE_TRACE();
+  // for NNG mapping no operation needed here
```
This method must be overwritten since it's an abstract method from the Mapping class but for Nearest Neighbor Mapping it's irrelevant or have I doing something wrong here?

DavidSCN replied:
Having the method here is fine. Only the comment does not apply here as it refers to the NNG mapping although the method itself is implemented in the base class, which might be confusing.

fsimonis commented on src/mapping/NearestNeighborBaseMapping.cpp:
```diff
@@ -0,0 +1,153 @@
+  if (distanceStatistics.empty()) {
+    PRECICE_INFO("Mapping distance not available due to empty partition.");
+  } else {
+    PRECICE_INFO("Mapping distance {}", distanceStatistics);
```
This prints min, max, avg and var of the distances.

fsimonis commented on src/mapping/NearestNeighborBaseMapping.hpp:
```diff
@@ -0,0 +1,71 @@
+  /// nn or nng
+  std::string MAPPING_NAME_SHORT;
+
+  mutable logging::Logger _log{"mapping::" + MAPPING_NAME};
```
All-caps names are used for constants. These aren't constants though.

fsimonis commented on src/mapping/NearestNeighborBaseMapping.hpp:
```diff
@@ -0,0 +1,76 @@
+  mutable logging::Logger _log{"mapping::" + MAPPING_NAME};
+
+  /// Compute the vector difference between the matched vector and the source vector (needed for gradient mapping)
+  std::vector<Eigen::VectorXd> _distancesMatched;
```
`Vertex::getCoords()` can return both 2d or 3d `Eigen::VectorXd`s. So this is fine.
If you feel the need for speed, you can save this as an `std::vector<bool>` and use an `Eigen::Map` to create an interface that uses the correct dimensions.

fsimonis commented on src/mapping/NearestNeighborGradientMapping.cpp:
```diff
@@ -0,0 +1,104 @@
+
+  if (hasConstraint(CONSERVATIVE)) {
+    PRECICE_DEBUG("Map conservative");
+    size_t const inSize = input()->vertices().size();
```
Clang-format should handle this.

DavidSCN replied:
AFAIK clang cannot handle this and the file here was not altered by the formatter. Why do you think clang-format takes care of this?

fsimonis reviewed (commented):
The gradient data is currently not accessible via the API, correct?

Some general notes on this PR:

* The `Mapping::MeshRequirement` is used to determine if connectivity information of a mesh is required. The comparison `operator<` is used to find the most specific requirement for all mapping requirements.
* The gradient information has nothing to do with this, so you should add it to `Mapping` as a boolean, which defaults to `false`. This way, you only need to specify it in your Gradient Mapping. Maybe you could name it `requireGradient`?
* `distanceMatched` contains offsets, not distances :wink: 
* There is still some logic in the nn-base-mapping that belongs into the gradient mapping. You can realize this by adding a `virtual void NearestNeighbourBaseMapping::onMappingComputed(const std::vector<size_t>& matchedVertices) {};`. You can then overwrite this in the NearestNeighbourGradientMapping and compute the `distancesMatched` there.
* Does the new mapping work/make sense with the scaled consistent feature? #906

ariguiba commented:
> The gradient data is currently not accessible via the API, correct?

Yes, the API Methods for the gradient would come in another Pull Requests when the reviews for this one are final (otherwise it would have been too hard to review).

> Some general notes on this PR:
> 
> * The `Mapping::MeshRequirement` is used to determine if connectivity information of a mesh is required. The comparison `operator<` is used to find the most specific requirement for all mapping requirements.
> * The gradient information has nothing to do with this, so you should add it to `Mapping` as a boolean, which defaults to `false`. This way, you only need to specify it in your Gradient Mapping. Maybe you could name it `requireGradient`?

Yes, that makes more sense. This is will be adjusted.

> * `distanceMatched` contains offsets, not distances wink

Yes, I imagine that is more accurate, thanks. 

> * There is still some logic in the nn-base-mapping that belongs into the gradient mapping. You can realize this by adding a `virtual void NearestNeighbourBaseMapping::onMappingComputed(const std::vector<size_t>& matchedVertices) {};`. You can then overwrite this in the NearestNeighbourGradientMapping and compute the `distancesMatched` there.

The problem with this option is that we would be calling a virtual method for every matched vector, which is what we wanted to avoid for performance reasons. The other option would be to save a list with all the matched vectors and iterate other these in the virtual method, but that is inefficient then in matter of storage. Is there another possibility maybe that you could suggest to go around these two problems or did i understand your idea wrong maybe ?

> * Does the new mapping work/make sense with the scaled consistent feature? #906

Not yet for now, but this could eventually be added. I just first need to look more into the details if it is at all possible.

fsimonis commented:
> The problem with this option is that we would be calling a virtual method for every matched vector, which is what we wanted to avoid for performance reasons.

My proposition is a hook that runs after the mapping was computed, not after each vertex mapping is computed.
On that note: the geometric point is called vertex, not vector.

> The other option would be to save a list with all the matched vectors and iterate other these in the virtual method, but that is inefficient then in matter of storage. Is there another possibility maybe that you could suggest to go around these two problems or did i understand your idea wrong maybe ?

This was my proposition. The list of all matched vertices already exists, so you can either pass it to the virtual function, or make the virtual function directly access it (like `NearestNeighbour::map()` does).
Iterating over vertices twice is not ideal, but also isn't an issue here.

uekerman commented:
This PR partly replaces #922

ariguiba commented:
## Update 

A little update to the changes added after the review: 
- Gradient removed as MeshRequirement (removed from < and = operators) and replaced as a boolean _requireGradient in the Mapping class
- Variables refactoring: distancesMatched -> offsetsMatched and meshDimensions -> spacialDimensions
- Added an onMapping-method to NearestNeighorGradient to calculate the offsets and removed this from NearestNeighborBase (so the offsets array is not unnecessarily allocated when using NearestNeighbor)
- Applied clang-format and added more logging information for errors and warnings

If I haven't overseen anything, these should be the major review suggestions/improvements. If everything is fine this way, I will base the second part of the PR, which would contain the API methods necessary to input gradient data, on this current state.

DavidSCN commented:
Before merging: could you add a changelog-entry?

DavidSCN commented on docs/changelog/1154.md:
```diff
@@ -0,0 +1,3 @@
+- Added a NearestNeighborBase class as a parent class for the NearestNeighbor and the NearestBeighborGradient mapping
+- Added NearestNeighborGradient Mapping, an additional mapping method, that uses gradient data input from the user
```
This sounds a bit like the mapping is already available in preCICE once this PR is merged. The changelog is not only for developers, but also for (interested) users. May I ask you to clarify a bit on this? Maybe write 'Added internals for NNG mapping implementation' or something similar.

ariguiba replied:
Yes I can do this :+1:
